### PR TITLE
[feat] trainer: add full training-state resume support

### PR DESCRIPTION
**Pull request by Reallkirn**
## Summary

This PR adds full training-state resume support for StarVLA training.

It separates two checkpoint-loading modes:

- `trainer.pretrained_checkpoint`: model-only warm start / fine-tuning
- `trainer.resume_from_checkpoint`: full training-state resume from a `steps_N_state` directory

Full resume restores:

- model state
- optimizer state
- scheduler state
- Accelerate / DeepSpeed state
- RNG state saved by Accelerate
- `completed_steps`
- native trainer dataloader cursor

Training-state checkpoints are saved atomically via:

```text
steps_N_state.tmp -> steps_N_state
```

This PR also guards distributed helper calls so single-process DeepSpeed smoke tests do not call `dist.get_rank()`, `dist.barrier()`, or `dist.destroy_process_group()` outside initialized process-group contexts.

## Validation

Validated on a single RTX 4090D with DeepSpeed ZeRO-2.

Config:

- framework: `QwenOFT`
- backbone: `Qwen/Qwen2.5-VL-3B-Instruct`
- dataset: LIBERO goal, `libero_goal`
- precision: bf16
- batch size: 1
- attention: `sdpa`
- `trainer.save_training_state=true`
- `trainer.save_model_checkpoint=false`

Validation flow:

1. Fresh training from step 0 to step 2 saved `steps_2_state`.
2. `resume_from_checkpoint=latest` restored `steps_2_state`.
3. Continued to step 4 and saved `steps_4_state`.
4. Explicit path resume from `steps_4_state`.
5. Continued to step 5 and saved `steps_5_state`.

Observed trainer states:

```text
steps_2_state: completed_steps=2, dataloader cursor=2
steps_4_state: completed_steps=4, dataloader cursor=4
steps_5_state: completed_steps=5, dataloader cursor=5
```

Consistency check from the same `steps_2_state`:

```text
Route A: step2 -> step4 -> resume -> step5
Route B: step2 -> directly train to step5
```

Results:

- model state SHA256 matched exactly
- optimizer tensor comparison passed with `max_diff = 0.0`
- no tensor key mismatch in optimizer states

This confirms that model parameters and optimizer tensor state are restored consistently across interrupted training.

## Reproduce

The validation used the same command pattern with different `max_train_steps` / `resume_from_checkpoint` values:

```bash
accelerate launch \
  --config_file starVLA/config/deepseeds/deepspeed_zero2.yaml \
  --num_processes 1 \
  --num_machines 1 \
  --mixed_precision bf16 \
  starVLA/training/train_starvla.py \
  --config_yaml examples/LIBERO/train_files/starvla_cotrain_libero.yaml \
  --framework.name QwenOFT \
  --framework.qwenvl.base_vlm ./playground/Pretrained_models/Qwen2.5-VL-3B-Instruct \
  --framework.qwenvl.attn_implementation sdpa \
  --framework.action_model.action_horizon 2 \
  --datasets.vla_data.data_root_dir playground/Datasets/LEROBOT_LIBERO_DATA \
  --datasets.vla_data.data_mix libero_goal \
  --datasets.vla_data.per_device_batch_size 1 \
  --trainer.freeze_modules qwen_vl_interface \
  --trainer.gradient_accumulation_steps 1 \
  --trainer.save_interval 2 \
  --trainer.eval_interval 999999 \
  --trainer.logging_frequency 1 \
  --trainer.save_training_state true \
  --trainer.save_model_checkpoint false \
  --run_root_dir ./playground/Checkpoints \
  --run_id resume_full_verify
```

Resume was tested with both:

```bash
--trainer.resume_from_checkpoint latest
```

and:

```bash
--trainer.resume_from_checkpoint ./playground/Checkpoints/resume_full_verify/checkpoints/steps_4_state
```

## Notes

Model-only loading remains available through `trainer.pretrained_checkpoint`.

Full training-state resume should use:

```bash
--trainer.resume_from_checkpoint latest
```

or:

```bash
--trainer.resume_from_checkpoint path/to/checkpoints/steps_N_state
```

A strict comparison was performed from the same `steps_2_state` because bitwise equivalence between two independent fresh runs is not guaranteed when model initialization occurs before seeding.

### CODE REVIEW

**Comment by JinhuiYE** [comment]
Sorry for the very late reply — it's been finals season with lots of deadlines piling up.

Thanks a lot for this great PR! The full training-state resume support is very helpful.

One question: did you implement it such that by default, saving the full training state is disabled? Since the state (optimizer, scheduler, DeepSpeed, etc.) takes a lot of storage space, we'd prefer not to enable it by default.

Users who need full resume functionality can explicitly opt in (e.g., via a flag like save_training_state=true). Could you confirm if this is already the case, or if any adjustment is needed?

Thanks again for your contribution!

**Comment by Reallkirn** [comment]
> Sorry for the very late reply — it's been finals season with lots of deadlines piling up.
> 
> Thanks a lot for this great PR! The full training-state resume support is very helpful.
> 
> One question: did you implement it such that by default, saving the full training state is disabled? Since the state (optimizer, scheduler, DeepSpeed, etc.) takes a lot of storage space, we'd prefer not to enable it by default.
> 
> Users who need full resume functionality can explicitly opt in (e.g., via a flag like save_training_state=true). Could you confirm if this is already the case, or if any adjustment is needed?
> 
> Thanks again for your contribution!

Wish you all the best with your work and studies! Thanks for checking this!

Yes, full training-state saving is disabled by default. Users have to explicitly opt in with `trainer.save_training_state=true`.

**I have thoroughly verified the three modified training entrypoints:**

- `starVLA/training/train_starvla.py`: `save_training_state` is read with default `False` around lines 302-303, and the full Accelerate/DeepSpeed state is only saved inside the `if save_training_state:` branch around line 327.
- `starVLA/training/train_starvla_cotrain.py`: same default `False` around lines 277-278, with full state saving gated by `if save_training_state:` around line 302.
- `starVLA/training/train_starvlm.py`: same default `False` around lines 262-263, with full state saving gated by `if save_training_state:` around line 287.

The existing lightweight/model-only checkpoint behavior remains enabled by default via `save_model_checkpoint=True`, so the default storage behavior should stay close to the previous behavior.

I also checked the configs/examples: there is no default training config enabling `save_training_state: true`; the only occurrence is in `docs/faq.md` as an explicit opt-in example.

Feel free to let me know if you'd like to rename the flag or revise the documentation.

**Comment by Reallkirn** [comment]
I synced this branch with the latest upstream `starVLA_dev` and resolved the conflict in `starVLA/training/train_starvla_cotrain.py`.

The conflict was around the eval condition in the co-training loop. I kept both behaviors:

- the upstream `optimizer_stepped` guard, so eval/save/log only happen after a real optimizer step
- the resume PR's `examples=batch_vla`, so evaluation does not consume an extra dataloader batch and affect the saved dataloader cursor

The resolved logic is:

`if optimizer_stepped and self.completed_steps % self.config.trainer.eval_interval == 0:
    step_metrics = self.eval_action_model(step_metrics, examples=batch_vla)`

I also reran basic checks locally:
- no unresolved conflict markers
- `git diff --check` passed
- related training files compile successfully with `compileall`

**Comment by yeahjack** [comment]
@codex review
